.github/workflows/ci.yaml

```
@@ -79,6 +79,86 @@ jobs:
         run: nox -R -e "$NOX_SESSION"
         timeout-minutes: 10
 
+  nox-cross-arch:
+    name: Cross-arch tests with nox
+    if: github.event_name == 'merge_group'
+    strategy:
+      fail-fast: false
+      # Before adding new items to this matrix, make sure that a dockerfile
+      # exists for the combination of items in the matrix.
+      # Refer to .github/containers/nox-cross-arch/README.md to learn how to
+      # add and name new dockerfiles.
+      matrix:
+        arch:
+          - arm64
+        os:
+          - ubuntu-20.04
+        python:
+          - "3.11"
+        nox-session:
+          - "pytest_min"
+    runs-on: ${{ matrix.os }}
+
+    steps:
+      - name: Fetch sources
+        uses: actions/checkout@v3
+
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v2
+        with:
+          platforms: linux/${{ matrix.arch }}
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+
+      # This is a workaround to prevent the cache from growing indefinitely.
+      # https://docs.docker.com/build/ci/github-actions/cache/#local-cache
+      # https://github.com/docker/build-push-action/issues/252
+      # https://github.com/moby/buildkit/issues/1896
+      - name: Cache container layers
+        uses: actions/cache@v3
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-buildx-nox-${{ matrix.arch }}-${{ matrix.os }}-${{ matrix.python }}
+
+      - name: Build image
+        uses: docker/build-push-action@v4
+        with:
+          context: .github/containers/nox-cross-arch
+          file: .github/containers/nox-cross-arch/${{ matrix.arch }}-${{ matrix.os }}-python-${{ matrix.python }}.Dockerfile
+          platforms: linux/${{ matrix.arch }}
+          tags: localhost/nox-cross-arch:latest
+          push: false
+          load: true
+          cache-from: type=local,src=/tmp/.buildx-cache
+          cache-to: type=local,dest=/tmp/.buildx-cache-new,mode=max
+
+      # Refer to the workaround mentioned above
+      - name: Move cache
+        run: |
+          rm -rf /tmp/.buildx-cache
+          mv /tmp/.buildx-cache-new /tmp/.buildx-cache
+
+      # Cache nox downloads
+      - name: Cache nox downloads
+        uses: actions/cache@v3
+        with:
+          path: |
+            .nox/${{ matrix.nox-session }}
+          key: nox-${{ matrix.nox-session }}-${{ matrix.arch }}-${{ matrix.os }}-${{ matrix.python }}-${{ hashFiles('**/pyproject.toml') }}
+
+      - name: Run nox
+        run: |
+          docker run \
+            --rm \
+            -v $(pwd):/${{ github.workspace }} \
+            -w ${{ github.workspace }} \
+            --net=host \
+            --platform linux/${{ matrix.arch }} \
+            localhost/nox-cross-arch:latest \
+            nox -R -e ${{ matrix.nox-session }}
```

The `-R` means to not create the `venv`, but this only makes sense if the `venv` was created before. Also for the CI we need to make sure the `venv` is (re)generated in case dependencies changed.

```suggestion
            nox -e ${{ matrix.nox-session }}
```

---

I was under the same impression, but I was wrong. In my tests, the env gets created if it does not exist.

---

Check this for an example: https://github.com/tiyash-basu-frequenz/frequenz-sdk-python/actions/runs/6082366114/job/16499995712

---

> Also for the CI we need to make sure the venv is (re)generated in case dependencies changed.

This requirement has also been addressed here. If the dependencies change, then `pyproject.toml` changes, which invalidates the cache. Then `nox` downloads everything all over again.

---

Really? The help says:

```
  -r, --reuse-existing-virtualenvs
                        Re-use existing virtualenvs instead of recreating them.
...
-R                    Re-use existing virtualenvs and skip package re-installation. This is an alias for '--reuse-existing-virtualenvs --no-install'.
...
  --no-install          Skip invocations of session methods for installing packages (session.install, session.conda_install, session.run_always) when a virtualenv is being reused.
```

In any case, we do really want to re-install on each run, to make sure we get the latest dependencies, at least for `pytest_max`. If we don't, I don't think dependencies will be upgraded, to do so `pip` needs to be invoked with `-U` and `nox` doesn't seem to do that.

---

> This requirement has also been addressed here. If the dependencies change, then `pyproject.toml` changes, which invalidates the cache. Then `nox` downloads everything all over again.

No, this is not exactly the case, we depend on `xxx >= 1.1.1, < 2` for example, then if the venv had installed 1.1.1 and then a new 1.2.0 is out, the package will not be upgraded by `nox` and the `pyproject.toml` file won't change either.

---

okay I'll remove `-R`.

---

Done. Also, added steps to cache pip downloads.